x/ccv/provider/keeper/permissionless.go

```diff
@@ -0,0 +1,440 @@
+		return errorsmod.Wrapf(types.ErrInvalidUpdateRecord,
+			"did not find update record for chain: %s", consumerId)
+	}
+
+	k.SetTopN(ctx, consumerId, updateRecord.Top_N)
+	k.SetValidatorsPowerCap(ctx, consumerId, updateRecord.ValidatorsPowerCap)
+	k.SetValidatorSetCap(ctx, consumerId, updateRecord.ValidatorSetCap)
+
+	k.DeleteAllowlist(ctx, consumerId)
+	for _, address := range updateRecord.Allowlist {
+		consAddr, err := sdk.ConsAddressFromBech32(address)
+		if err != nil {
+			continue
+		}
+
+		k.SetAllowlist(ctx, consumerId, types.NewProviderConsAddress(consAddr))
+	}
+
+	k.DeleteDenylist(ctx, consumerId)
+	for _, address := range updateRecord.Denylist {
+		consAddr, err := sdk.ConsAddressFromBech32(address)
+		if err != nil {
+			continue
+		}
+
+		k.SetDenylist(ctx, consumerId, types.NewProviderConsAddress(consAddr))
+	}
+
+	oldTopN, found := k.GetTopN(ctx, consumerId)
+	if !found {
+		oldTopN = 0
+		k.Logger(ctx).Info("consumer chain top N not found, treating as 0", "consumerId", consumerId)
+	}
+
+	// if the top N changes, we need to update the new minimum power in top N
+	if updateRecord.Top_N != oldTopN {
+		if updateRecord.Top_N > 0 {
+			// if the chain receives a non-zero top N value, store the minimum power in the top N
+			bondedValidators, err := k.GetLastBondedValidators(ctx)
+			if err != nil {
+				return err
+			}
+			minPower, err := k.ComputeMinPowerInTopN(ctx, bondedValidators, updateRecord.Top_N)
+			if err != nil {
+				return err
+			}
+			k.SetMinimumPowerInTopN(ctx, consumerId, minPower)
+		} else {
+			// if the chain receives a zero top N value, we delete the min power
+			k.DeleteMinimumPowerInTopN(ctx, consumerId)
+		}
+	}
+
+	k.SetMinStake(ctx, consumerId, updateRecord.MinStake)
+	k.SetInactiveValidatorsAllowed(ctx, consumerId, updateRecord.AllowInactiveVals)
+
+	return nil
+}
+
+// GetLaunchedConsumersReadyToStop returns the consumer ids of the pending launched consumer chains
+// that are ready to stop
+func (k Keeper) GetLaunchedConsumersReadyToStop(ctx sdk.Context) []string {
+	store := ctx.KVStore(k.storeKey)
+	iterator := storetypes.KVStorePrefixIterator(store, types.ConsumerIdToStopTimeKeyNamePrefix())
+	defer iterator.Close()
+
+	var consumerIds []string
+
+	for ; iterator.Valid(); iterator.Next() {
+		// the `consumerId` resides in the whole key, but we skip the first byte (because it's the `ConsumerIdKey`)
+		// plus 8 more bytes for the `uint64` in the key that contains the length of the `consumerId`
+		consumerId := string(iterator.Key()[1+8:])
+
+		var stopTime time.Time
+		err := stopTime.UnmarshalBinary(iterator.Value())
+		if err != nil {
+			panic(fmt.Errorf("failed to unmarshal stop stopTime: %w for consumer id: %s", err, consumerId))
+		}
+
+		// if current block time is equal to or after stop stopTime, and the chain is launched we can stop the chain
+		phase, found := k.GetConsumerIdToPhase(ctx, consumerId)
+		if !ctx.BlockTime().Before(stopTime) && (found && phase == Launched) {
```

Comment:
You could potentially stop an initialized chain to prevent it from launching. But, not sure, I see a benefit in this: i) a consumer chain can just decide to not consume ICS security if they do not want to launch, and ii) they can leave the chain to fail to launch (e.g., if no validator opts in), iii) if they really want to prevent a launch (i.e., consumer client being created) they can set the `spawnTime` to be 100 years from now.

Reply:
The benefit is not having extra logic for only allowing launched consumers to be stopped. Also, allowing all consumers to be stopped it's an easy way to do garbage collection.